platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/sources/CustomGeometrySource.java

```diff
@@ -0,0 +1,176 @@
+package com.mapbox.mapboxsdk.style.sources;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.UiThread;
+import android.support.annotation.WorkerThread;
+
+import com.mapbox.mapboxsdk.geometry.LatLngBounds;
+import com.mapbox.mapboxsdk.style.layers.Filter;
+import com.mapbox.services.commons.geojson.Feature;
+import com.mapbox.services.commons.geojson.FeatureCollection;
+
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Custom Vector Source, allows using FeatureCollections.
+ *
+ */
+@UiThread
+public class CustomGeometrySource extends Source {
+  private ThreadPoolExecutor executor;
+  private GeometryTileProvider provider;
+  private Map<TileID, AtomicBoolean> cancelledTileRequests;
+
+  /**
+   * Create a CustomGeometrySource
+   *
+   * @param id the source id
+   */
+  public CustomGeometrySource(String id, GeometryTileProvider provider_) {
+    executor = new ThreadPoolExecutor(2, 2, 60, TimeUnit.SECONDS, new ArrayBlockingQueue(80));
```

Comment:
Missing generic classification on `ArrayBlockingQueue`: `<Runnable>`

Reply:
Should we ensure proper shutdown of the executor when the source is removed btw? At the moment, the core source might be removed while there are still tasks in the queue, these keep executing until finalisation of the Java object.